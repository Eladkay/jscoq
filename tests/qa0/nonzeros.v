Require Import List.
Import ListNotations.


Definition natlist := list nat.

Fixpoint nonzeros (l : natlist) :=
  match l with
  | [] => []
  | 0 :: xs => nonzeros xs
  | x :: xs => x :: nonzeros xs
  end.

Check nonzeros.
Compute nonzeros [5;6;0;7;0;9;8].

Arguments nonzeros l : simpl nomatch.

Lemma nonzeros_app : forall l1 l2 : natlist,
  nonzeros (l1 ++ l2) = (nonzeros l1) ++ (nonzeros l2).
Proof.
  intros l1 l2. induction l1 as [| n' l' IHl'].
  -simpl. reflexivity.
  -simpl. case n'; simpl.
   + assumption.
   + congruence.
Qed.

Check nonzeros_app.
