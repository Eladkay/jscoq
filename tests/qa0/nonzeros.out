------------------------------------------------------------
[Notice] nonzeros
              : natlist -> list nat
------------------------------------------------------------
------------------------------------------------------------
[Notice]      = [5; 6; 7; 9; 8]
              : list nat
------------------------------------------------------------
------------------------------------------------------------
[Notice] nonzeros_app
              : forall l1 l2 : natlist, nonzeros (l1 ++ l2) = nonzeros l1 ++ nonzeros l2
------------------------------------------------------------
